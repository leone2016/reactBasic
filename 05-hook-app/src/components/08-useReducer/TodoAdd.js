import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });
    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }
        handleAddTodo(newTodo)
        reset();
    }


    return (
        <form onSubmit={handleSubmit} >
            <input
                type="text"
                name="description"
                className="from-control"
                placeholder="Aprender ..."
                autoComplete="off"
                value={description}
                onChange={handleInputChange}
            />
            <button type="submit " class="btn btn-info" >Agregar TODO</button>
        </form>
    )
}
