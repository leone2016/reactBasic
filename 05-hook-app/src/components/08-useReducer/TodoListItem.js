import React from 'react'

export const TodoListItem = ({ todo,index, handleDelete, handleToggle }) => {
    return (
        <li key={todo.id} className="">
            <span onClick={() => handleToggle(todo.id)}
                className={`${todo.done && 'complete'}`} >
                {index + 1}) {todo.desc}
            </span>
            <button class="btn btn-danger"
                onClick={() => handleDelete(todo.id)}>
                Borrar
            </button>
        </li>

    )
}
