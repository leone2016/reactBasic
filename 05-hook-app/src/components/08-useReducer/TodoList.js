import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul>
            {
                todos.map((todo, index) => (
                    <TodoListItem
                        key={todo.id}
                        todo={todo}
                        index={index}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                ))
            }
        </ul>
    )
}
