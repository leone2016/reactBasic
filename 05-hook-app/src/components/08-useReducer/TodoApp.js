import React, { useReducer, useEffect, useCallback } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todo')) || [];
    /* return [{
        id: new Date().getTime(),
        desc: 'Aprender react',
        done: false
    
    }] */
}
export const TodoApp = () => {


    /**
     * recibe 3 parametros
     * useReducer: funcion reducer
     * initialState: estado inicial
     * init: funcion que se utilizar para inicializar el State en caso de que el state 
     *      sea procesado o haga varias acciones, cada que el componente se vuelve a redibujar
     * Ã² ejecutar esa accion, entonces al ser una funcion externa al componente lo hace mas ligero y rapido
     * incluso puede memorizar ese resultado en caso de necesitarlo
     */
    /*  const [state, dispatch] = useReducer(useReducer, initialState, init) */
    const [todos, dispatch] = useReducer(todoReducer, [], init);
 

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

   
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }
    const handleDelete = useCallback(
        (todoId) => {
            const action = {
                type: 'delete',
                payload: todoId
            }
            dispatch(action);
        },
        [],
    )

    const handleToggle = useCallback(
        (todoId) => {
            dispatch({
                type: 'toggle',
                payload: todoId
            });
        },
        [],
    )

    return (
        <div>
            <h2>TodoApp ({todos.length})</h2>

            <TodoAdd handleAddTodo={handleAddTodo}/>
            <hr />
            <TodoList
                todos={todos}
                handleDelete={handleDelete}
                handleToggle={handleToggle}
            />

        </div>
    )
}
