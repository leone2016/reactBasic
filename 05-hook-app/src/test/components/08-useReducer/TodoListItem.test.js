import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

//clase 154
describe('Prueba en <TodoListItem/>', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            index={0}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
        />
    )
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('debe de llamar la funcion handleBorrar', () => {
        wrapper.find('button').simulate('click');
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id);

    });
    test('debe de llamar la funcion handletoggle', () => {

        wrapper.find('span').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id);

    });
    test('debe de llamar la funcion correctamente', () => {

        const p = wrapper.find('span');
        expect(p.text().trim()).toBe(`1) ${demoTodos[0].desc}`);
    });
    test('debe de tener la clase complete si el TODO.done  = true ', () => {

        const todo = demoTodos[0];
        todo.done = true;
        const wrapper = shallow(
            <TodoListItem
                todo={todo}
            />
        )
        expect(wrapper.find('span').hasClass('complete')).toBe(true);
    });
}) 