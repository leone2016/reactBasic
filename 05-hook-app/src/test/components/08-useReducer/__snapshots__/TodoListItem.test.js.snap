// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <TodoListItem/> debe de mostrarse correctamente 1`] = `
<li
  className=""
  key="1"
>
  <span
    className="false"
    onClick={[Function]}
  >
    1
    ) 
    Aprender React
  </span>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
